provider/machine_pool_resource.go

```diff
@@ -464,6 +554,86 @@ func (r *MachinePoolResource) Update(ctx context.Context, request tfsdk.UpdateRe
+		return false, fmt.Errorf("failed to get information for cluster %s: %v", state.Cluster.Value, err)
+	}
+	cluster := resp.Body()
+	isMultiAZCluster := cluster.MultiAZ()
+	clusterAZs := cluster.Nodes().AvailabilityZones()
+	clusterSubnets := cluster.AWS().SubnetIDs()
+
+	if isMultiAZCluster {
+		// Can't set both availability_zone and subnet_id
+		if !common.IsStringAttributeEmpty(state.AvailabilityZone) && !common.IsStringAttributeEmpty(state.SubnetID) {
+			return false, fmt.Errorf("availability_zone and subnet_id are mutually exclusive")
+		}
+
+		// multi_availability_zone setting must be consistent with availability_zone and subnet_id
+		azOrSubnet := !common.IsStringAttributeEmpty(state.AvailabilityZone) || !common.IsStringAttributeEmpty(state.SubnetID)
+		if !state.MultiAvailabilityZone.Null && !state.MultiAvailabilityZone.Unknown {
+			if azOrSubnet && state.MultiAvailabilityZone.Value {
+				return false, fmt.Errorf("multi_availability_zone must be False when availability_zone or subnet_id is set")
+			}
+		} else {
+			state.MultiAvailabilityZone = types.Bool{Value: !azOrSubnet}
+		}
+	} else { // not a multi-AZ cluster
+		if !common.IsStringAttributeEmpty(state.AvailabilityZone) {
+			return false, fmt.Errorf("availability_zone can only be set for multi-AZ clusters")
+		}
+		if !common.IsStringAttributeEmpty(state.SubnetID) {
+			return false, fmt.Errorf("subnet_id can only be set for multi-AZ clusters")
+		}
+		if !state.MultiAvailabilityZone.Null && !state.MultiAvailabilityZone.Unknown && state.MultiAvailabilityZone.Value {
+			return false, fmt.Errorf("multi_availability_zone can only be set for multi-AZ clusters")
+		}
+		state.MultiAvailabilityZone = types.Bool{Value: false}
+	}
+
+	// Ensure that the machine pool's AZ and subnet are valid for the cluster
+	if !common.IsStringAttributeEmpty(state.AvailabilityZone) {
+		inClusterAZ := false
+		for _, az := range clusterAZs {
+			if az == state.AvailabilityZone.Value {
+				inClusterAZ = true
+				break
+			}
+		}
+		if !inClusterAZ {
+			return false, fmt.Errorf("availability_zone %s is not valid for cluster %s", state.AvailabilityZone.Value, state.Cluster.Value)
+		}
+	} else {
+		if len(clusterAZs) > 0 && !state.MultiAvailabilityZone.Value && isMultiAZCluster {
+			state.AvailabilityZone = types.String{Value: clusterAZs[0]}
+		} else {
+			state.AvailabilityZone = types.String{Null: true}
+		}
+	}
+	if !common.IsStringAttributeEmpty(state.SubnetID) {
+		inClusterSubnet := false
+		for _, subnet := range clusterSubnets {
+			if subnet == state.SubnetID.Value {
+				inClusterSubnet = true
+				break
+			}
+		}
+		if !inClusterSubnet {
+			return false, fmt.Errorf("subnet_id %s is not valid for cluster %s", state.SubnetID.Value, state.Cluster.Value)
+		}
+	} else {
+		if len(clusterSubnets) > 0 && !state.MultiAvailabilityZone.Value && isMultiAZCluster {
+			state.SubnetID = types.String{Value: clusterSubnets[0]}
+		} else {
+			state.SubnetID = types.String{Null: true}
+		}
+	}
```

Comment:
If I understand the code right here, there might be situation in which you set both az and subnet, am I right?
So it possible that user sets az and code use `clusterSubnets[0]`?
If I understand it right from the cli code, if user ask to singleAZ but doesn't set either AZ or subnet, it takes only `clusterAZs[0]` and keep subnet empty

Reply:
done